[chart.js] Use RadialLinearScale type for scale option

indigolain opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/chartjs/Chart.js/blob/6632b8ba8475e020fd0533e2630117376a9cbc73/src/controllers/controller.radar.js#L13
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

indigolain commented:
This change is required to enable the radar chart to use the ticks option and more.
The file linked in the pull request description defines the scale type of the radar chart to be radialLinear, which is defined below and is already defined in the current index.d.ts .
https://github.com/chartjs/Chart.js/blob/87e44fa360be580dae085d85d74614dfdc3988c5/src/scales/scale.radialLinear.js

related: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/37614

indigolain commented:
@adripanico 
Thank you for the review!
I also checked the document, and I believe that what the document is saying in the quoted text is that it takes the singular `scale` and not the plural `scales` option.
> The radar chart supports only a single scale. The options for this scale are defined in the scale property.

I've made a codepen to test out if I can use the `radialLinear` scale options, and it seems to work. It'll be great if you could check it out.
https://codepen.io/indigolain/pen/XWrZydB

indigolain commented:
I looked into the documentation of Chart.js deeper and found an example where scale options other than `display` was used, will this make this pull request approvable?

https://www.chartjs.org/docs/latest/axes/radial/linear.html#axis-range-settings

adripanico commented on types/chart.js/index.d.ts:
```diff
@@ -663,6 +664,7 @@ declare namespace Chart {
     }
 
     interface RadialLinearScale extends LinearScale {
+        display?: boolean;
```
RadialLinearScale extends LinearScale which extends ChartScales which already define the ```display``` property.

indigolain commented:
I think I can use the RadialChartOptions interface, introduced in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26688 , when defining the option, and then use the option to get the desired result.
Since you could also use this interface for specifying the display option, should we revert the change introduced in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/37614 ?

ElianCordoba commented:
Okay so I just try every configuration that the `scale` have with your change and I found the following:

```
// These are defined in the RadialLinearScale interface
lineArc:     don't know what it does but in your code pen does not change anithing
angleLines:  works
pointLabels  works
ticks:       works
// This come from ChartScales which itself comes from LinearScale
type:        don't work
scaleLabel:  don't work
position:    don't work
type:        don't work
display      works
gridLines    works
xAxes        don't work
yAxes        don't work
```

My fix would be to replate the current type with 
```typescript
// Note I'm not extending from LinearScale
interface RadialLinearScale {
  lineArc?: boolean; // Need confirmation on this one
  angleLines?: AngleLineOptions;
  pointLabels?: PointLabelOptions;
  ticks?: TickOptions;
  display?: boolean;
  gridLines?: GridLineOptions;
}
```

indigolain commented:
@ElianCordoba 
Thank you for checking out the configuration! From looking at the chart.js implementation [here](https://github.com/chartjs/Chart.js/blob/87e44fa360be580dae085d85d74614dfdc3988c5/src/scales/scale.radialLinear.js), I confirmed that the configs below, which includes the configs you suggested, are valid.
- display
- animate (This may not be working  ref: https://github.com/chartjs/Chart.js/issues/5749)
- position
- angleLines
- gridLines
- ticks
- pointLabels

Also, it seems that `lineArc` config was replaced with angleLines, gridLines, and pointLabels config in the pull request below, so I think we should remove it if we don't have to think about 
backward compatibility.
https://github.com/chartjs/Chart.js/pull/3782

@adripanico 
The change @ElianCordoba suggested will narrow down the invalid config that a user may specify, but I believe that this will not solve the problem you mentioned about the generic options interface used openly. What do you think about this?

indigolain commented:
I've edited the types based on @ElianCordoba 's idea, and added some tests for the changes that I made.

And I've also added some type definitions based on the references below.

For PositionType
https://github.com/chartjs/Chart.js/blob/db6a6d5989cb33dda38c3ecbe2ce0b4b19b3d013/src/core/core.layouts.js#L208

For GridLineOptions
https://www.chartjs.org/docs/latest/axes/styling.html#grid-line-configuration

For PointLabelOptions
https://www.chartjs.org/docs/latest/axes/radial/linear.html#point-label-options

ElianCordoba commented:
Looks good to me, @adripanico what do you think?

indigolain commented:
@adripanico  @ElianCordoba 
Thank you for reviewing my codes!

adripanico commented:
> @adripanico @ElianCordoba
> Thank you for reviewing my codes!

Thanks to you!
